File: app/src/test/java/org/aossie/agoraandroid/SignUpTest.java
package org.aossie.agoraandroid;

import android.os.Build;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.aossie.agoraandroid.remote.APIService;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.io.IOException;

import okhttp3.OkHttpClient;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.converter.scalars.ScalarsConverterFactory;

@Config(sdk = Build.VERSION_CODES.O_MR1)
@RunWith(RobolectricTestRunner.class)
public class SignUpTest {
    private MockWebServer mockWebServer = new MockWebServer();
    private APIService apiService;

    @Before
    public void setup() throws IOException {
        Gson gson = new GsonBuilder().setLenient().create();
        mockWebServer.start();
        apiService = new Retrofit.Builder().baseUrl(mockWebServer.url("/")).client(new OkHttpClient()).addConverterFactory(ScalarsConverterFactory.create()).addConverterFactory(GsonConverterFactory.create(gson)).build().create(APIService.class);
    }

    @After
    public void teardown() throws IOException {
        mockWebServer.shutdown();
    }

    @Test
    public void signUpTest() throws JSONException, IOException {
        //Body for post request
        JSONObject jsonObject = new JSONObject();
        JSONObject securityJsonObject = new JSONObject();
        jsonObject.put("identifier", "fakePersonName");
        jsonObject.put("password", "fakePassword");
        jsonObject.put("email", "fakeEmail@gmail.com");
        jsonObject.put("firstName", "fakeFirstName");
        jsonObject.put("lastName", "fakeLastName");
        securityJsonObject.put("question", "fakeQuestion");
        securityJsonObject.put("answer", "fakeAnswer");
        jsonObject.put("securityQuestion",securityJsonObject);
        //Response Body
        JSONObject responseBack = new JSONObject();
        responseBack.put("token","Some token string");
        responseBack.put("expiresOn","Some date time or timestamp");
        MockResponse mockResponse = new MockResponse().setResponseCode(200).setBody(responseBack.toString());
        mockWebServer.enqueue(mockResponse);
        Response responseFromRequest = apiService.createUser(jsonObject.toString()).execute();
        Assert.assertEquals(responseFromRequest.body(),responseBack.toString());
        MockResponse mockUserNameExists = new MockResponse().setResponseCode(409);
        mockWebServer.enqueue(mockUserNameExists);
        responseFromRequest = apiService.createUser(jsonObject.toString()).execute();
        Assert.assertEquals(responseFromRequest.code(),409);
    }
}

File: app/src/test/java/org/aossie/agoraandroid/apitesting/authentication/SignUpTest.kt
package org.aossie.agoraandroid.apitesting.authentication

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import okhttp3.mockwebserver.MockResponse
import org.aossie.agoraandroid.apitesting.BaseTest
import org.aossie.agoraandroid.data.remote.dto.NewUserDto
import org.aossie.agoraandroid.data.remote.dto.NewUserDtoJsonAdapter
import org.aossie.agoraandroid.common.utilities.MockFileParser
import org.junit.Assert
import org.junit.Test
import retrofit2.Response
import java.io.IOException

@ExperimentalCoroutinesApi
class SignUpTest : BaseTest() {

  @Test
  @Throws(IOException::class)
  fun signUpTest() {

    val signUpRequest: NewUserDto? = NewUserDtoJsonAdapter(moshi).fromJson(MockFileParser("requests/auth_requests/signUp_request.json").content)
    val signUpResponse: String = MockFileParser("responses/auth_responses/signUp_response.json").content

    signUpRequest?.let {
      mockWebServer.enqueue(MockResponse().setBody(signUpResponse))
      testScope.launch {
        val responseFromRequest: Response<*> = apiService.createUser(it)
        Assert.assertEquals(responseFromRequest.body(), signUpResponse)
      }
    }
  }
}
